File: C++/142.cpp
__________________________________________________________________________________________________
8ms
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
static const auto speedup = []() {std::ios::sync_with_stdio(false); std::cin.tie(NULL); return 0; }();

class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        if(head == nullptr) return nullptr;
        ListNode* slow = head;
        ListNode* fast = head;
        bool findCycle = false;
        while(slow && fast && fast->next)
        {
            slow = slow->next;
            fast = fast->next->next;
            if(slow == fast) 
            {
                findCycle = true;
                break;
            }
        }
        if(!findCycle) return nullptr;
        
        slow = head;
        while(slow != fast)
        {
            slow = slow->next;
            fast = fast->next;
        }
        return slow;
    }
};
__________________________________________________________________________________________________
9548 kb
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        bool cycle = 0;
        ListNode *runner1 = head;
        ListNode *runner2 = head;
        if(head == NULL) return {};
        while(runner2->next != NULL && runner2->next->next != NULL) {
            runner1 = runner1->next;
            runner2 = runner2->next->next;
            if(runner1 == runner2) {
                cycle = 1;
                break;
            }
        }
        if(cycle) {
            runner1 = head;
            while(runner1 != runner2) {
                runner1 = runner1->next;
                runner2 = runner2->next;
            }
            return runner1;
        }
        return {};
    }
};
__________________________________________________________________________________________________

File: Java/142.java
__________________________________________________________________________________________________
sample 0 ms submission
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        if (null == head || null == head.next)
            return null;
        ListNode p, q;
        p = head.next;
        q = head.next.next;
        while (q != null && p != q) {
            p = p.next;
            if (q.next == null)
                q = null;
            else 
                q = q.next.next;
        }
        if (q == null)
            return null;
        p = head;
        while (p != q) {
            p = p.next;
            q = q.next;
        }
        return p;
    }
}
__________________________________________________________________________________________________
sample 31436 kb submission
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        ListNode root = new ListNode(0);
        root.next = head;
        ListNode p1 = root;
        ListNode p2 = root;
        ListNode p;
        while (true) {
            if (p1 != null && p1.next != null && p1.next.next != null) {
                p1 = p1.next.next;
                p2 = p2.next;
            } else {
                
                System.gc();
                return null;
            }
            if (p1 == p2) {
                p = p1;
                break;
            }
        }
        ListNode h = root;
        while (true) {
            h = h.next;
            p = p.next;
            if (h == p) {
                System.gc();
                return h;
            }
        }
    }
}
__________________________________________________________________________________________________
